kong/router/atc.lua

```diff
@@ -186,32 +198,32 @@ end
 
 local function categorize_fields(fields)
 
-  if not is_http then
-    return fields, nil, nil
-  end
-
   local basic = {}
-  local headers = {}
-  local queries = {}
 
   -- 13 bytes, same len for "http.queries."
   local PREFIX_LEN = 13 -- #"http.headers."
 
   for _, field in ipairs(fields) do
     local prefix = field:sub(1, PREFIX_LEN)
 
-    if prefix == "http.headers." then
-      headers[field:sub(PREFIX_LEN + 1)] = field
+    if prefix == HTTP_HEADERS_PREFIX then
+      if not basic[HTTP_HEADERS_PREFIX] then
+        basic[HTTP_HEADERS_PREFIX] = {}
+      end
+      table_insert(basic[HTTP_HEADERS_PREFIX], field:sub(PREFIX_LEN + 1))
```

Comment:
This is rather slow, we can store the field callback like this:

```
local CALLBACKS = {
  ["http.method"] =
    function(v, params, cb)
      return cb("http.method", params.method)
    end,

    ["http.path"] =
    function(v, params, cb)
      return cb("http.path", params.uri)
    end,

    ["http.host"] =
    function(v, params, cb)
      return cb("http.host", params.host)
    end,
    ...
}

setmetatable(CALLBACKS, {
  __index = function(k)
    if k:sub(1, 13) == "http.headers." then
      return function()
        return headers[k:sub(14)]
      end
    end
  end
})
```

Or something like this.

Reply:
done.